### PR TITLE
Refactor classes related to Python AST for clarity and consistency

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Renamed ast_helpers.py to python_ast.py
Renamed ASTLinter to Tree, Visitor to TreeVisitor
Merged AstHelper into Tree
Extracted list_import_sources into ImportSourcesCollector
Extracted list_sys_path_changes into SysPathChangesCollector

### Linked issues
None

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

**Review by asnare — APPROVED** [review]

File: src/databricks/labs/ucx/source_code/graph.py [hunk 1]
```diff
@@ -164,10 +163,10 @@ def visit(self, visit_node: Callable[[DependencyGraph], bool | None], visited: s
 
     def build_graph_from_python_source(self, python_code: str) -> list[DependencyProblem]:
         problems: list[DependencyProblem] = []
-        linter = ASTLinter.parse(python_code)
-        syspath_changes = DbutilsLinter.list_sys_path_changes(linter)
-        run_calls = DbutilsLinter.list_dbutils_notebook_run_calls(linter)
-        import_sources, import_problems = DbutilsLinter.list_import_sources(linter, DependencyProblem)
+        tree = Tree.parse(python_code)
+        syspath_changes = SysPathChangesCollector.collect_sys_path_changes(tree)
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/graph.py:167]:
```suggestion
        syspath_changes = SysPathChange.extract_from(tree)
```

File: src/databricks/labs/ucx/source_code/graph.py [hunk 2]
```diff
@@ -164,10 +163,10 @@ def visit(self, visit_node: Callable[[DependencyGraph], bool | None], visited: s
...
+        run_calls = DbutilsLinter.list_dbutils_notebook_run_calls(tree)
+        import_sources, import_problems = ImportSourcesCollector.collect_import_sources(tree, DependencyProblem)
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/graph.py:169]:
```suggestion
        import_sources, import_problems = ImportSource.extract_from(tree, DependencyProblem)
```

File: src/databricks/labs/ucx/source_code/linters/pyspark.py [hunk 3]
```diff
@@ -328,24 +327,24 @@ def name(self) -> str:
         return self._from_table.name()
 
     def lint(self, code: str) -> Iterable[Advice]:
-        linter = ASTLinter.parse(code)
-        for node in TreeWalker.walk(linter.root):
+        tree = Tree.parse(code)
+        for node in TreeWalker.walk(tree.root):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/pyspark.py:331]:
```suggestion
        for node in tree.walk():
```
Technically, walk can be __iter__

File: src/databricks/labs/ucx/source_code/linters/python_ast.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+logger = logging.getLogger(__file__)
+
+missing_handlers: set[str] = set()
+
+
+T = TypeVar("T", bound=NodeNG)
+
+
+class Tree(Generic[T]):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/python_ast.py:18]:
```suggestion
class Tree:
```

File: src/databricks/labs/ucx/source_code/linters/python_ast.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    @property
+    def root(self):
+        return self._root
+
+    def locate(self, node_type: type[T], match_nodes: list[tuple[str, type]]) -> list[T]:
+        visitor = MatchingVisitor(node_type, match_nodes)
+        visitor.visit(self._root)
+        return visitor.matched_nodes
+
+    def collect_sys_paths_changes(self):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/python_ast.py:37]: Are we sure we want this coupling here?

File: src/databricks/labs/ucx/source_code/linters/python_ast.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        if callable(method_slot):
+            method_slot(node)
+        else:
+            self.visit_nodeng(node)
+
+    def visit_nodeng(self, node: NodeNG):
+        pass
+
+
+class TreeWalker:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/python_ast.py:147]: Remove

File: src/databricks/labs/ucx/source_code/linters/imports.py [hunk 7]
```diff
@@ -1,298 +1,25 @@
...
-class AbsolutePath(SysPathChange):
-    pass
-
-
-# path added to sys.path using os.path.abspath
-class RelativePath(SysPathChange):
-    pass
-
-
-class SysPathVisitor(Visitor):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/imports.py:149]: This visitor is related to imports, so has to stay in this module

**Review by nfx — CHANGES_REQUESTED** [review]
See extended comments in https://github.com/databrickslabs/ucx/pull/1850

**Comment by ericvergnaud** [comment]
@nfs ready for review

**Review by nfx — APPROVED** [review]
Lgtm

**Pull request merged**